### PR TITLE
CM-831 - Broker Architectural Pattern in C#

### PR DESCRIPTION
**Pull request by muhammedsalimp**
CM-831 -Broker Architectural Pattern in C#

### CODE REVIEW

File: csharp-architectural-patterns/BrokerPattern/EventBroker/Broker.cs [hunk 1]
```diff
@@ -0,0 +1,30 @@
+﻿using EventBroker.Interfaces;
+
+namespace EventBroker
```
Comment (mahrncic) [inline csharp-architectural-patterns/BrokerPattern/EventBroker/Broker.cs:3]: Let's use file-scoped namespaces everywhere

File: csharp-architectural-patterns/BrokerPattern/EventBroker/EventBroker.csproj [hunk 2]
```diff
@@ -0,0 +1,10 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <OutputType>Exe</OutputType>
+    <TargetFramework>net7.0</TargetFramework>
```
Comment (mahrncic) [inline csharp-architectural-patterns/BrokerPattern/EventBroker/EventBroker.csproj:5]: .NET 8 is out, so let's use it 😄

File: csharp-architectural-patterns/BrokerPattern/EventBroker/Interfaces/IBroker.cs [hunk 3]
```diff
@@ -0,0 +1,8 @@
+﻿namespace EventBroker.Interfaces
```
Comment (mahrncic) [inline csharp-architectural-patterns/BrokerPattern/EventBroker/Interfaces/IBroker.cs:1]: file-scoped

File: csharp-architectural-patterns/BrokerPattern/EventBroker/Message.cs [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+    {
+        public Message(string topic, string data)
+        {
+            Topic = topic ?? throw new ArgumentNullException(nameof(topic));
+            Data = data ?? throw new ArgumentNullException(nameof(data));
+        }
+
+        public string Topic { get; set; }
+        public string Data { get; set; }
+
```
Comment (mahrncic) [inline csharp-architectural-patterns/BrokerPattern/EventBroker/Message.cs:13]: Remove this blank line

File: csharp-architectural-patterns/BrokerPattern/MessagePublisher1/MessagePublisher1.cs [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+            await Task.Delay(5000);
+            await SendMessageAsync();
+            await Task.Delay(5000);
+            await SendMessageAsync();
+        }
+
+        private async Task SendMessageAsync()
+        {
+            DateTime currentDateTime = DateTime.Now;
+            string message = new($"Message from sender for topic:{topicName} at {currentDateTime}!");
```
Comment (mahrncic) [inline csharp-architectural-patterns/BrokerPattern/MessagePublisher1/MessagePublisher1.cs:27]: Use `var`

File: csharp-architectural-patterns/BrokerPattern/MessagePublisher1/MessagePublisher1.csproj [hunk 6]
```diff
@@ -0,0 +1,19 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <OutputType>Exe</OutputType>
+    <TargetFramework>net7.0</TargetFramework>
```
Comment (mahrncic) [inline csharp-architectural-patterns/BrokerPattern/MessagePublisher1/MessagePublisher1.csproj:5]: .NET 8

File: csharp-architectural-patterns/BrokerPattern/MessagePublisher1/Program.cs [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Hosting;
+
+namespace MessagePublisherApp1
+{
+    public static class Program
+    {
+        public static async Task Main(string[] args)
+        {
+            HostApplicationBuilder builder = Host.CreateApplicationBuilder();
```
Comment (mahrncic) [inline csharp-architectural-patterns/BrokerPattern/MessagePublisher1/Program.cs:11]: `var`

File: csharp-architectural-patterns/BrokerPattern/MessagePublisher2/MessagePublisher2.cs [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+            await Task.Delay(5000);
+            await SendMessageAsync();
+            await Task.Delay(5000);
+            await SendMessageAsync();
+        }
+
+        private async Task SendMessageAsync()
+        {
+            DateTime currentDateTime = DateTime.Now;
+            string message = new($"Message from sender for topic:{topicName} at {currentDateTime}!");
```
Comment (mahrncic) [inline csharp-architectural-patterns/BrokerPattern/MessagePublisher2/MessagePublisher2.cs:27]: `var`

File: csharp-architectural-patterns/BrokerPattern/MessagePublisher2/MessagePublisher2.csproj [hunk 9]
```diff
@@ -0,0 +1,19 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <OutputType>Exe</OutputType>
+    <TargetFramework>net7.0</TargetFramework>
```
Comment (mahrncic) [inline csharp-architectural-patterns/BrokerPattern/MessagePublisher2/MessagePublisher2.csproj:5]: .NET 8

File: csharp-architectural-patterns/BrokerPattern/MessagePublisher2/Program.cs [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+
+namespace MessagePublisherApp2
+{
+    public static class Program
+    {
+
+        public static async Task Main(string[] args)
+        {
+
+            HostApplicationBuilder builder = Host.CreateApplicationBuilder();
```
Comment (mahrncic) [inline csharp-architectural-patterns/BrokerPattern/MessagePublisher2/Program.cs:13]: `var`

File: csharp-architectural-patterns/BrokerPattern/MessagePublisherSubscriberTests/MessagePublisherSubscriberTests.csproj [hunk 11]
```diff
@@ -0,0 +1,34 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFramework>net7.0</TargetFramework>
```
Comment (mahrncic) [inline csharp-architectural-patterns/BrokerPattern/MessagePublisherSubscriberTests/MessagePublisherSubscriberTests.csproj:4]: .NET 8

File: csharp-architectural-patterns/BrokerPattern/MessageReceiver/MessageReceiverApp.csproj [hunk 12]
```diff
@@ -0,0 +1,18 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <OutputType>Exe</OutputType>
+    <TargetFramework>net7.0</TargetFramework>
```
Comment (mahrncic) [inline csharp-architectural-patterns/BrokerPattern/MessageReceiver/MessageReceiverApp.csproj:5]: .NET 8

File: csharp-architectural-patterns/BrokerPattern/MessageReceiver/Program.cs [hunk 13]
```diff
@@ -0,0 +1,21 @@
...
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Hosting;
+
+namespace MessageReceiverApp
+{
+    public static class Program
+    {
+        public static async Task Main(string[] args)
+        {
+            HostApplicationBuilder builder = Host.CreateApplicationBuilder();
```
Comment (mahrncic) [inline csharp-architectural-patterns/BrokerPattern/MessageReceiver/Program.cs:11]: `var`

File: csharp-architectural-patterns/BrokerPattern/MessageSender/MessageSender.cs [hunk 14]
```diff
@@ -0,0 +1,31 @@
...
+            await Task.Delay(5000);
+            await SendMessageAsync();
+            await Task.Delay(5000);
+            await SendMessageAsync();
+        }
+
+        private async Task SendMessageAsync()
+        {
+            DateTime currentDateTime = DateTime.Now;
+            string message = new($"Message from sender at {currentDateTime}!");
```
Comment (mahrncic) [inline csharp-architectural-patterns/BrokerPattern/MessageSender/MessageSender.cs:26]: `var`

File: csharp-architectural-patterns/BrokerPattern/MessageSender/MessageSenderApp.csproj [hunk 15]
```diff
@@ -0,0 +1,18 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <OutputType>Exe</OutputType>
+    <TargetFramework>net7.0</TargetFramework>
```
Comment (mahrncic) [inline csharp-architectural-patterns/BrokerPattern/MessageSender/MessageSenderApp.csproj:5]: .NET 8

File: csharp-architectural-patterns/BrokerPattern/MessageSender/Program.cs [hunk 16]
```diff
@@ -0,0 +1,21 @@
...
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Hosting;
+
+namespace MessageSenderApp
+{
+    public class Program
+    {
+        public static async Task Main()
+        {
+            HostApplicationBuilder builder = Host.CreateApplicationBuilder();
```
Comment (mahrncic) [inline csharp-architectural-patterns/BrokerPattern/MessageSender/Program.cs:11]: `var`

File: csharp-architectural-patterns/BrokerPattern/MessagingService/MessageService.cs [hunk 17]
```diff
@@ -0,0 +1,116 @@
...
+        }
+
+        private static ServiceBusClient CreateClient()
+        {
+            var clientOptions = new ServiceBusClientOptions()
+            {
+                TransportType = ServiceBusTransportType.AmqpWebSockets
+            };
+            ServiceBusClient? client = new(connectionString, clientOptions);
+            return client;
```
Comment (mahrncic) [inline csharp-architectural-patterns/BrokerPattern/MessagingService/MessageService.cs:86]: The `return` statement should always have a blank line before it. Make sure to apply it where possible.

File: csharp-architectural-patterns/BrokerPattern/MessagingService/MessageService.cs [hunk 18]
```diff
@@ -0,0 +1,116 @@
...
+
+            string body = args.Message.Body.ToString();
+            messageHandlerCallback(body);
+            await args.CompleteMessageAsync(args.Message);
+        }
+
+        private Task ErrorHandler(ProcessErrorEventArgs args)
+        {
+            Console.WriteLine(args.Exception.ToString());
+            return Task.CompletedTask;
```
Comment (mahrncic) [inline csharp-architectural-patterns/BrokerPattern/MessagingService/MessageService.cs:113]: Blank line before `return`

**Review by mahrncic — CHANGES_REQUESTED** [review]
Great job in overall, a couple of changes are needed on most of the files and projects:

- use .NET 8
- use file-scoped namespaces
- use `var` where possible
- always add a blank line before a `return` statement where possible

File: csharp-architectural-patterns/BrokerPattern/MessagePublisher2/Program.cs [hunk 19]
```diff
@@ -0,0 +1,22 @@
...
+using Microsoft.Extensions.Hosting;
+
+namespace MessagePublisherApp2;
+
+public static class Program
+{
+
+    public static async Task Main(string[] args)
+    {
+
```
Comment (mahrncic) [inline csharp-architectural-patterns/BrokerPattern/MessagePublisher2/Program.cs:12]: This blank line should be removed

File: csharp-architectural-patterns/BrokerPattern/MessagePublisherSubscriberTests/MessagePublisherSubscriberTests.cs [hunk 20]
```diff
@@ -0,0 +1,101 @@
+using MessagePublisherApp1;
+using MessagePublisherApp2;
+using MessageService;
+using MessageSubscriberApp1;
+using MessageSubscriberApp2;
+using MessageSubscriberApp3;
+using Moq;
+
+namespace MessagePublisherSubscriberTests
+{
```
Comment (mahrncic) [inline csharp-architectural-patterns/BrokerPattern/MessagePublisherSubscriberTests/MessagePublisherSubscriberTests.cs:10]: File-scoped namespace should be used

File: csharp-architectural-patterns/BrokerPattern/MessageSubscriber3/Program.cs [hunk 21]
```diff
@@ -0,0 +1,21 @@
+﻿using MessageService;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Hosting;
+
+namespace MessageSubscriberApp3;
+
+public static class Program
+{
+
```
Comment (mahrncic) [inline csharp-architectural-patterns/BrokerPattern/MessageSubscriber3/Program.cs:9]: Remove this blank line

**Review by mahrncic — CHANGES_REQUESTED** [review]
A couple of blank lines and one non file-scoped namespace left

**Review by mahrncic — APPROVED** [review]

**Review by mahrncic — APPROVED** [review]

File: csharp-architectural-patterns/BrokerPattern/MessagingService/MessageService.cs [hunk 22]
```diff
@@ -0,0 +1,76 @@
+﻿using Azure.Messaging.ServiceBus;
+
+namespace MessageService;
+
+public class MessageService : IMessageService
+{
+    Action<string>? messageHandlerCallback;
+    private static readonly string connectionString = "{service bus connection string}";
+
+    static ServiceBusClient ServiceBusClient => CreateClient();
```
Comment (jshergal) [inline csharp-architectural-patterns/BrokerPattern/MessagingService/MessageService.cs:10]: This should be `private static readonly` and the `=>` should be just `=`. As it is written now we are still creating a new client for every access of `ServiceBusClient`

**Pull request closed without merging**